### ISSUE TITLE
[FEAT] arm64 build

### ISSUE THREAD

**Issue by sOmEoNe2lOvEgIt** [eythaann/Seelen-UI#321]
**Is your feature request related to a problem? Please describe.**
The app is currently not supported on ARM64 devices, it currently runs on the prism emulation layer.
Even if it is realy well optimised, it would greately benefit from an ARM64 build especially as they are becoming more and more common (partly with the copilot plus pc's).

**Describe the solution you'd like**
Publish releases compiled for ARM64 windows computers.

**Describe alternatives you've considered**
Keep using the app under the emulation layer.

**Additional context**
The latest copilot pc's have the benefit of having a greater battery life than prior models mostly due to the adoption of the arm architecture by microsoft and other manufacturers. They will become more and more present and relevent as time goes on so providing an arm build would most certainly, in long term, benefit to the project.

**Comment by Keima-Katsuragi** [comment]
Yes, I think it is not difficult to construct it.  [https://learn.microsoft.com/zh-cn/windows/arm/overview](url)

**Comment by eythaann** [comment]
is not difficult, but I need an ARM processor to test.

**Comment by sOmEoNe2lOvEgIt** [comment]
Might be a stretch but i could test the compiled app for you on my laptop (surface pro 11, non-oled so slower snapdragon processor but shouldn't matter i guess).

**Comment by eythaann** [comment]
yeah the hard works is not compile it, the problem is debug it, if you're a dev  and you wanna work on this, you're welcome.

**Comment by vld-b** [comment]
Hello eythaann. I am currently trying to compile the app for arm64 in order to try and work on it and get it working on arm64 in general, but I am facing an issue. I see you've built the app on the Tauri framework, therefore I just tried building it with `cargo build` and `cargo build --release`. None worked. I always get the same error with the same `ring v0.17.14` package, saying that it can't find the "assert.h" file. I've downloaded rustup and all visual studio components correctly. Kindly tell me more if you know about this error and maybe you also encountered it when building on x64.

**Comment by eythaann** [comment]
no idea.

**Comment by teapot-spun08** [comment]
Are there still plans for an ARM native version? Does anyone know how well it currently runs on Prism emulation?

**Comment by eythaann** [comment]
yes I want to add support but I need to buy a laptop with snapdragon.

**Comment by vld-b** [comment]
I tried to build Seelen-UI for ARM64, and while I successfully rewrote the SIMD instructions to NEON intrinsics, there's a memory leak somewhere in the communication between the background service and the frontend, but I don't understand the program's architecture enough to fix that. I used a Surface Pro 11 with X Plus.

**Comment by eythaann** [comment]
the memory leak is an issue on tauri/webview2 https://github.com/tauri-apps/tauri/issues/4026 https://github.com/tauri-apps/tauri/issues/9190

**Comment by eythaann** [comment]
but if you have a branch that successfully compiles and runs on ARM64 that could be great to get merged.

**Comment by vld-b** [comment]
My results up until now are that it compiles (after a long time of fighting with the ring module), and the background service runs ok, but the fronted just allocates gigabytes upon gigabytes of memory without even showing anything, then crashes and is restarted by the backend. Also thanks for the links to the issues, but they don't seem to be arm64-specific. Both report leaks from x64 macOS and Windows respectively.

**Comment by eythaann** [comment]
Oh I thought you were referring to an already reported memory leak after some time of usage, your case looks like another thing.

**Comment by vld-b** [comment]
to be fair the memory leak I'm observing seems to be coming from a `std::fs::read_to_string`, a `TcpStream::connect` or a `serde_json::to_writer`, but I don't know for sure. Maybe you can make something out of the flamegraph?

<img width="1865" height="788" alt="Image" src="https://github.com/user-attachments/assets/a4def9f2-acd8-4ef7-bab6-035ec30c224e" />

**Comment by eythaann** [comment]
Hmm I really don't have any idea what could be. That's why I have to buy an arm laptop to make this.

**Comment by vld-b** [comment]
Makes sense lol

**Comment by screech** [comment]
I have a branch I had created a while ago where I had seen this issue (allocating gigabytes and then crashing) and worked around it. Let me see if I can revive the branch. OTOH if you have a current ARM branch with the NEON code that I can use .. happy to try and patch that up

**Comment by vld-b** [comment]
Absolutely! If you found the fix for the memory leak I'd be happy to share my code with the NEON intrinsics.

**Comment by screech** [comment]
I have to try with a current tree .. can you point me to your tree ?

**Comment by vld-b** [comment]
My code is in https://github.com/vld-b/Seelen-UI-arm64/tree/feature/arm64 ; The only changes are to the .gitignore, src/background/utils/icon_extractor/mod.rs (NEON intrinsics) and src/service/cli/tcp.rs (function rewrite).

**Comment by screech** [comment]
Ok .. I see the same issue again.

this is what I had had to hack out

use windows::Win32::{
    Foundation::{BOOL, HWND, LPARAM},
    UI::WindowsAndMessaging::{EnumChildWindows, EnumWindows},
};

use crate::error::Result;

#[derive(Debug, Clone)]
pub struct WindowEnumerator {
    parent: Option<HWND>,
}

impl WindowEnumerator {
    pub fn new() -> Self {
        Self { parent: None }
    }

The problem was the win32 api which seemed to not work for aarch64 .. in
case someone knows why

**Comment by vld-b** [comment]
So you got it working now? Also, in which file are you seeing this?

**Comment by screech** [comment]
not yet .. porting from my old version now.

The code has moved a lot

**Comment by screech** [comment]
so .. just took this patch, applied to TOT and it just worked for me..

diff --git a/src/background/utils/icon_extractor/mod.rs
b/src/background/utils/icon_extractor/mod.rs
index a9d9bca4..79258007 100644
--- a/src/background/utils/icon_extractor/mod.rs
+++ b/src/background/utils/icon_extractor/mod.rs
@@ -18,10 +18,18 @@ use windows::Win32::{

 use seelen_core::state::Icon;

+#[cfg(target_arch = "x86_64")]
 use std::arch::x86_64::{
     __m128i, _mm_loadu_si128, _mm_setr_epi8, _mm_shuffle_epi8,
_mm_storeu_si128,
 };
+
+#[cfg(target_arch = "aarch64")]
+use std::arch::aarch64::{
+    uint8x16_t, vld1q_u8, vqtbl1q_u8, vst1q_u8
+};
+
 use std::io::BufRead;
+use std::os::windows::ffi::OsStrExt;
 use std::path::{Path, PathBuf};

 use crate::error::Result;
@@ -37,6 +45,7 @@ use crate::windows_api::WindowsApi;
 /// Convert BGRA to RGBA
 ///
 /// Uses SIMD to go fast
+#[cfg(target_arch = "x86_64")]
 pub fn bgra_to_rgba(data: &mut [u8]) {
     // The shuffle mask for converting BGRA -> RGBA
     let mask: __m128i = unsafe {
@@ -55,6 +64,26 @@ pub fn bgra_to_rgba(data: &mut [u8]) {
     }
 }

+// Uses NEON intrinsics to go fast
+#[cfg(target_arch = "aarch64")]
+pub fn bgra_to_rgba(data: &mut [u8]) {
+    // The shuffle mask for converting BGRA -> RGBA
+    let maskplain: [u8; 16] = [
+        2, 1, 0, 3, // First pixel
+        6, 5, 4, 7, // Second pixel
+        10, 9, 8, 11, // Third pixel
+        14, 13, 12, 15 // Fourth pixel
+    ];
+    // The shuffle mask for the conversion in NEON intrinsics
+    let mask: uint8x16_t = unsafe { vld1q_u8(maskplain.as_ptr()) };
+    // For each 16-byte chunk in your data
+    for chunk in data.chunks_exact_mut(16) {
+        let mut vector: uint8x16_t = unsafe { vld1q_u8(chunk.as_ptr()) };
+        vector = unsafe { vqtbl1q_u8(vector, mask) };
+        unsafe { vst1q_u8(chunk.as_mut_ptr(), vector) };
+    }
+}
+
 pub fn convert_hicon_to_rgba_image(hicon: &HICON) -> Result<RgbaImage> {
     unsafe {

**Comment by vld-b** [comment]
What exactly worked for you? And on what architecture?

**Comment by screech** [comment]
Everything ... it did not do the "allocating GB's" which happened when I
used your tree.

This is a aarch64 Surface 7 .. and I can use the entire suite with native
compilation (have to check if any of the libraries are still cross compiled)

./target/debug/seelen-ui.exe
./target/debug/seelen_ui.pdb
./target/debug/static/themes/animated-start-icon/seelen
./target/debug/static/themes/default/seelen
./target/debug/static/themes/dock-animation/seelen

***@***.*** CLANGARM64 ~/seelenui/Seelen-UI-orig (arm64)
$ file ./target/debug/seelen-ui.exe
./target/debug/seelen-ui.exe: PE32+ executable (console) Aarch64, for MS
Windows, 9 sections

**Comment by vld-b** [comment]
This sounds amazing! Does this mean it's ready for a PR and to finally include arm64 build?

**Comment by screech** [comment]
I am building the optimized build now and will send that as a PR

for the current TOT for the "dev" build, I just took your changes for NEON
into TOT .. if you want to try that and resubmit if it works for you too,
then I can send the changes I needed for a "release" build

**Comment by vld-b** [comment]
Hi can you point me to your tree/codebase you used that worked for you so I can clone it to try on my surface pro 11? Also which surface did you use? Pro or Laptop?

**Comment by screech** [comment]
> Hi can you point me to your tree/codebase you used that worked for you so I can clone it to try on my surface pro 11? Also which surface did you use? Pro or Laptop?

https://github.com/screech/Seelen-UI/commits/Sep25-arm/

Look for the last two commits on TOT .. should be all you need

**Comment by screech** [comment]
with those .. I can even build/install an optimized aarch64 build. @eythaann let me know if you want to see a PR for these

**Comment by eythaann** [comment]
you have to test the build step, execute `npm run tauri build`, test the installer and the app for at least 1 week to see if all is ok.

**Comment by vld-b** [comment]
I couldn't compile it successfully because of an error in the `ring v0.17.14` module. It's the module that gave me problems last time as well. @screech mind telling me what rustc toolchain you're using, or perhaps you know the fix directly?

Edit: my error specifically is something about failed to execute custom build command for `ring`, failed to find source file "assert.h"

**Comment by screech** [comment]
Don't know of an issue with ring.

This is what I am using

$ rustup toolchain list
stable-aarch64-pc-windows-msvc (default)
nightly-2024-06-25-aarch64-pc-windows-msvc
nightly-aarch64-pc-windows-msvc (override)

**Comment by screech** [comment]
$ rustup show
Default host: aarch64-pc-windows-msvc
rustup home:  C:\Users\shrij\.rustup

installed toolchains
--------------------

stable-aarch64-pc-windows-msvc (default)
nightly-2024-06-25-aarch64-pc-windows-msvc
nightly-aarch64-pc-windows-msvc

active toolchain
----------------

nightly-aarch64-pc-windows-msvc (overridden by
'C:\Users\shrij\seelenui\Seelen-UI\rust-toolchain.toml')
rustc 1.91.0-nightly (7aef4bec4 2025-09-01)

**Comment by vld-b** [comment]
Nvm I managed to fix this error by using VS2022 native tools command prompt. I'll report shortly on how the program works

**Comment by vld-b** [comment]
Yay I got it working on my surface pro 11. I got this error at the end of building: "A public key has been found, but no private key. Make sure to set `TAURI_SIGNING_PRIVATE_KEY` environment variable.
       Error A public key has been found, but no private key. Make sure to set `TAURI_SIGNING_PRIVATE_KEY` environment variable."
Other than that, I installed Seelen UI on my SP11. The installer was still x86, but the program itself that displays the UI and the slu-service are both arm64. I'll be testing the program on the tablet over the course of the next couple of weeks and especially when school starts again and I'll report back.
So good news arm64 support will be coming soon!!!

**Comment by vld-b** [comment]
Week one report:
Seelen-UI in itself works well on ARM64. These tests were conducted with a plugged-in Microsoft Surface Pro 11-
Pros:
- Works smoothly on ARM64 architecture (most of the time)
- Supports most if not all features correctly that x64 Seelen-UI does

Cons:
- Doesn't support many plugins (maybe has something to do with the fact that this isn't an official build). It usually says something like "This plugin was built for an older version of Seelen" or similar stuff. Even if I download them and activate them, some to most don't work
- Doesn't handle going to sleep well. Sometimes after waking up Seelen-UI crashes and comes back shortly, other times slu-service AND UI crash and Seelen-UI has to be restarted manually, sometimes (usually if the device hasn't slept for too long) Seelen-UI works normally
- Is fairly heavy. Normal usage of device showed that it was heating up when it should not (e.g. browsing with Edge), which wasn't happening without Seelen-UI

This was all I could observe. One honourable con was that I went to a sleepover one night and the day after when I came back home, my Surface Pro 11 was extremely hot to the touch (the hottest it had been) and the fan was running on max speed (without anybody having used the Surface Pro 11). I didn't know how long this had gone on for, but this didn't happen before Seelen-UI. I'm not saying that this was because of Seelen-UI, but it might have something to do with this.

Starting next week I will be conducting tests on battery and report back.

**Comment by screech** [comment]
>
>
>    - Doesn't handle going to sleep well. Sometimes after waking up
>    Seelen-UI crashes and comes back shortly, other times slu-service AND UI
>    crash and Seelen-UI has to be restarted manually, sometimes (usually if the
>    device hasn't slept for too long) Seelen-UI works normally
>
>
I have definitely seen this .. and after syncing with TOT I am now seeing
the dock just not showing up

>
>    - Is fairly heavy. Normal usage of device showed that it was heating
>    up when it should not (e.g. browsing with Edge), which wasn't happening
>    without Seelen-UI
>
> This I did not see .. Seelen seemed to run pretty light weight on my
Surface 7.

**Comment by eythaann** [comment]
@screech that are bugs present on last release, being fixed for next release.

**Comment by eythaann** [comment]
@vld-b thanks for making the branch, you can make an PR to check the changes please?

**Comment by vld-b** [comment]
@eythaann It is indeed not me who has the branch but @screech. I built from https://github.com/screech/Seelen-UI/tree/Sep25-arm and @screech just copied my arm64 rust code with the NEON intrinsics from https://github.com/vld-b/Seelen-UI-arm64/commit/6e418ca8be7516e766eece7118f06f4b2d78bdf5 and pasted it into his branch. Wouldn't he have to make the PR in this case?

**Comment by screech** [comment]
I can send a PR this weekend and we can merge and re-try

**Comment by eythaann** [comment]
@screech well, only is needed that function change? and it will compile successfully? what about the memory leaks? also that branch is outdated @vld-b

**Comment by vld-b** [comment]
My branch is outdated. @screech 's is up to date. I can also update my branch, compile and see if Seelen UI works the same as from @screech 's branch and PR from my branch.

**Comment by vld-b** [comment]
I got impatient and PR'd https://github.com/eythaann/Seelen-UI/pull/1179

**Comment by vld-b** [comment]
2-Week report of Seelen-UI on SP11 arm64:
This week's testing was mostly done on battery, thus the battery saving Minimal and Extreme were tested.
Pros:
- Features worked on battery like they worked plugged in
- Battery saving modes worked

Cons:
- Slightly higher battery consumption (noticeable over a 7-hour day), but it didn't drain the battery completely
- UI elements don't interact with touch and only partially interact with pen (with Slim Pen 2 a double tap acted the same way as a normal mouse click, usually)

This is about what I could make out. If you have any questions feel free to ask!

**Comment by eythaann** [comment]
TouchScreen will be fully supported on v2.5 as an entire refactor on main widgets like toolbar and dock will be done.

**Comment by vld-b** [comment]
Hey @eythaann have you already checked out my PR https://github.com/eythaann/Seelen-UI/pull/1179 ? It's the PR that contains the NEON intrinsics for the arm64 build, if you want to release arm64.

**Comment by eythaann** [comment]
<img width="943" height="629" alt="Image" src="https://github.com/user-attachments/assets/4ea8a372-2ff3-48a5-b411-2dcd76561211" />

Ok after fighting a little with the github actions, I got multi-arch working on the nightly build.

**Comment by eythaann** [comment]
v2.4.4 release is done, but I will wait for SignPath to give me the OSS sign code lincense for production. (no more screen warning finally)

**Comment by vld-b** [comment]
Lets gooo 🥳
